crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -1,14 +1,404 @@
-use std::marker::PhantomData;
+use {
+    crate::{
+        component::{
+            func::{self, LiftContext, LowerContext},
+            matching::InstanceType,
+            values::{ErrorContextAny, FutureAny, StreamAny},
+            Val,
+        },
+        vm::component::{
+            ComponentInstance, LocalErrorContextRefCount, StateTable, StreamFutureState,
+            WaitableState,
+        },
+        AsContextMut,
+    },
+    anyhow::{bail, Result},
+    std::{marker::PhantomData, mem::MaybeUninit},
+    wasmtime_environ::component::{
+        CanonicalAbiInfo, InterfaceType, TypeFutureTableIndex, TypeStreamTableIndex,
+    },
+};
+
+#[derive(Copy, Clone, Debug)]
+enum TableIndex {
+    Stream(TypeStreamTableIndex),
+    Future(TypeFutureTableIndex),
+}
+
+fn state_table(instance: &mut ComponentInstance, ty: TableIndex) -> &mut StateTable<WaitableState> {
+    let runtime_instance = match ty {
+        TableIndex::Stream(ty) => instance.component_types()[ty].instance,
+        TableIndex::Future(ty) => instance.component_types()[ty].instance,
+    };
+    &mut instance.component_waitable_tables()[runtime_instance]
+}
+
+fn get_mut_by_index_from(
+    state_table: &mut StateTable<WaitableState>,
+    ty: TableIndex,
+    index: u32,
+) -> Result<(u32, &mut StreamFutureState)> {
+    Ok(match ty {
+        TableIndex::Stream(ty) => {
+            let (rep, WaitableState::Stream(actual_ty, state)) =
+                state_table.get_mut_by_index(index)?
+            else {
+                bail!("invalid stream handle");
+            };
+            if *actual_ty != ty {
+                bail!("invalid stream handle");
+            }
+            (rep, state)
+        }
+        TableIndex::Future(ty) => {
+            let (rep, WaitableState::Future(actual_ty, state)) =
+                state_table.get_mut_by_index(index)?
+            else {
+                bail!("invalid future handle");
+            };
+            if *actual_ty != ty {
+                bail!("invalid future handle");
+            }
+            (rep, state)
+        }
+    })
+}
+
+fn host_close_reader<U, S: AsContextMut<Data = U>>(store: S, rep: u32) -> Result<()> {
```

Comment:
I'm not sure how much this will persist int he future but you can drop the `<Data = U>` here in the sense that it's not necessary as part of this definition. In the future if you need bounds on `U` it should be possible to do `S::Data: Bound` I think too

Reply:
Good catch.  This is just a stub, with the signature copied from https://github.com/dicej/wasmtime/blob/d1596df0361a4a8c956ba7ef34ac8d3c5e0140e1/crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs#L578, but it looks like the real version doesn't use `U` either, so I'm guessing I had originally been using `U`, then stopped using it, but forgot to update the signature.